libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,363 @@
+				)
+				{
+					ops.swap(ops.sourceSize() - offset - 1);
+					return true;
+				}
+
+		// If a lower slot should be removed, try to bring up the slot that should end up there and bring it up.
+		// Note that after the cases above, there will always be a target slot to duplicate in this case.
+		for (size_t offset: ranges::views::iota(0u, ops.sourceSize()))
+			if (
+				!ops.isCompatible(offset, offset) && // The lower slot is not already in position.
+				ops.sourceMultiplicity(offset) < 0 && // We have too many copies of this slot.
+				offset <= ops.targetSize() && // There is a target slot at this position.
+				!ops.targetIsArbitrary(offset) // And that target slot is not arbitrary.
+			)
+			{
+				bringUpTargetSlot(ops, offset);
+				return true;
+			}
+
+		// At this point we want to keep all slots.
+		for (size_t i = 0; i < ops.sourceSize(); ++i)
+			yulAssert(ops.sourceMultiplicity(i) >= 0, "");
+		yulAssert(ops.sourceSize() <= ops.targetSize(), "");
+
+		// If the top is not in position, try to find a slot that wants to be at the top and swap it up.
+		if (!ops.isCompatible(sourceTop, sourceTop))
+			for (size_t sourceOffset: ranges::views::iota(0u, ops.sourceSize()))
+				if (
+					!ops.isCompatible(sourceOffset, sourceOffset) &&
+					ops.isCompatible(sourceOffset, sourceTop)
+				)
+				{
+					ops.swap(ops.sourceSize() - sourceOffset - 1);
+					return true;
+				}
+
+		// If we still need more slots, produce a suitable one.
+		if (ops.sourceSize() < ops.targetSize())
+		{
+			bringUpTargetSlot(ops, ops.sourceSize());
+			return true;
+		}
+
+		// The stack has the correct size, each slot has the correct number of copies and the top is in position.
+		yulAssert(ops.sourceSize() == ops.targetSize(), "");
+		size_t size = ops.sourceSize();
+		for (size_t i = 0; i < ops.sourceSize(); ++i)
+			yulAssert(ops.sourceMultiplicity(i) == 0 && (ops.targetIsArbitrary(i) || ops.targetMultiplicity(i) == 0), "");
+		yulAssert(ops.isCompatible(sourceTop, sourceTop), "");
+
+		// If we find a lower slot that is out of position, but also compatible with the top, swap that up.
+		for (size_t offset: ranges::views::iota(0u, size))
+			if (!ops.isCompatible(offset, offset) && ops.isCompatible(sourceTop, offset))
+			{
+				ops.swap(size - offset - 1);
+				return true;
+			}
+		// Swap up any slot that is still out of position.
+		for (size_t offset: ranges::views::iota(0u, size))
+			if (!ops.isCompatible(offset, offset) && !ops.sourceIsSame(offset, sourceTop))
+			{
+				ops.swap(size - offset - 1);
+				return true;
+			}
+		yulAssert(false, "");
+	}
+};
+
+
+/// Transforms @a _currentStack to @a _targetStack, invoking the provided shuffling operations.
+/// Modifies @a _currentStack itself after each invocation of the shuffling operations.
+template<typename Swap, typename PushOrDup, typename Pop>
+void createStackLayout(Stack& _currentStack, Stack const& _targetStack, Swap _swap, PushOrDup _pushOrDup, Pop _pop)
+{
+	struct ShuffleOperations
```

Comment:
Is this essentially the same as in `createIdealLayout`?

Reply:
Essentially: yes - but due to the symbolic ``PreviousSlot``s it differs everywhere slightly (hence abstracting the operations in the first place).